dex-it-common/src/main/scala/com/wavesplatform/dex/it/docker/PortBindingKeeper.scala

```diff
@@ -0,0 +1,81 @@
+package com.wavesplatform.dex.it.docker
+
+import cats.implicits._
+import com.github.dockerjava.api.command.CreateContainerCmd
+import com.github.dockerjava.api.model.Ports.Binding
+import com.github.dockerjava.api.model.{ExposedPort, PortBinding}
+import com.wavesplatform.dex.domain.utils.ScorexLogging
+
+import java.net.ServerSocket
+import java.util
+import scala.jdk.CollectionConverters._
+import scala.util.{Random, Using}
+
+object PortBindingKeeper extends ScorexLogging {
+
+  case class AvailablePortsNotFound() extends Throwable
+
+  private val stringRange = Option(System.getenv("TEST_PORT_RANGE"))
+    .getOrElse(throw new RuntimeException("Please specify the TEST_PORT_RANGE environment variable"))
+
+  private val portsRange = parsePortRange()
+  private var currentPosition = portsRange.start
+
+  def getBindings(exposedPorts: Seq[Int]): util.List[PortBinding] = this.synchronized {
+    log.debug(s"Getting ports for $exposedPorts")
+    findFreePorts(exposedPorts.size).map { result =>
+      exposedPorts.map(new ExposedPort(_)).zip(result.map(Binding.bindPort)).map { ports =>
+        new PortBinding(ports._2, ports._1)
+      }
+    } match {
+      case Right(v) =>
+        log.debug("Successfully generated ports")
+        v.asJava
+      case Left(ex) => throw ex
+    }
+  }
+
+  def getBindings(cmd: CreateContainerCmd, exposedPorts: Seq[Int]): util.List[PortBinding] = {
+    val cmdExposedPorts = cmd.getExposedPorts.toList.map(_.getPort)
+    PortBindingKeeper.getBindings(cmdExposedPorts ++ exposedPorts)
+  }
+
+  def getBindings(exposedPort: Int): util.List[PortBinding] =
+    getBindings(Seq(exposedPort))
+
+  private def findFreePorts(portNumbers: Int): Either[AvailablePortsNotFound, Seq[Int]] =
+    (0 to portNumbers).foldLeft(Seq.empty[Int].asRight[AvailablePortsNotFound]) { (acc, _) =>
+      acc.flatMap { accValue =>
+        findFreePort().map(_ +: accValue)
+      }
+    }
+
+  def findFreePort(): Either[AvailablePortsNotFound, Int] = {
+    val result = (currentPosition to portsRange.end).find { i =>
+      Using(new ServerSocket(i))(_.getLocalPort).isSuccess
+    }.orElse {
+      (portsRange.start to currentPosition).find { i =>
+        Using(new ServerSocket(i))(_.getLocalPort).isSuccess
+      }
+    }
```

Comment:
Also it looks like we don't need to scan ports in `[currentPosition; portsRange.end]` in `orElse` case, because we scanned them. Probably it should be:
```scala
val result = findFreePortIn(currentPosition to portsRange.end)
  .orElse(findFreePortIn(portsRange.start until currentPosition))
```

Reply:
I think, we may never check ports from the start of the range, because we provide range for every test, so there may never be that at some point we released ports and we need to take them again.